### PR TITLE
257 notifications using storefront api

### PR DESCRIPTION
**Pull request by kostyrin**

### CODE REVIEW

File: Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Core/Model/Search/NotificationMessageSearchCriteria.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    public class NotificationMessageSearchCriteria : SearchCriteriaBase
+    {
+        /// <summary>
+        /// 
+        /// </summary>
+        public bool IsActive { get; set; }
+        /// <summary>
+        /// time interval used to evaluate  active notifications have failure delivery
+        /// </summary>
+        public int RepeatHoursIntervalForFail { get; set; }
```
Comment (tatarincev) [inline Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Core/Model/Search/NotificationMessageSearchCriteria.cs]: Why we can't use Start and End date instead? Use RepeatHoursIntervalForFail  reduce the flexibility of usage and complicate the level of data access due to mixing business and data access logic?

File: Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Data/Services/NotificationMessageService.cs [hunk 2]
```diff
@@ -19,19 +18,25 @@ public class NotificationMessageService : INotificationMessageService
...
+            _notificationService = notificationService;
         }
 
         public async Task<NotificationMessage[]> GetNotificationsMessageByIds(string[] ids)
         {
             using (var repository = _repositoryFactory())
             {
                 var messages = await repository.GetMessagesByIdsAsync(ids);
-                return messages.Select(n => n.ToModel(AbstractTypeFactory<NotificationMessage>.TryCreateInstance())).ToArray();
+                return messages.Select(n =>
```
Comment (tatarincev) [inline Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Data/Services/NotificationMessageService.cs]: Why we can load related  notifications on the repository level as navigation properties with further converting into domain types in entity models?

File: Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Data/Services/NotificationSearchService.cs [hunk 3]
```diff
@@ -68,6 +69,8 @@ public async Task<NotificationSearchResult> SearchNotificationsAsync(Notificatio
                 var allPersistentProvidersTypes = result.Results.Select(x => x.GetType()).Distinct();
                 transientNotificationsQuery = transientNotificationsQuery.Where(x => !allPersistentProvidersTypes.Contains(x.GetType()));
 
+                transientNotificationsQuery = transientNotificationsQuery.Where(x => !x.Kind.EqualsInvariant(x.Type));
```
Comment (tatarincev) [inline Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Data/Services/NotificationSearchService.cs]: Why we need this check?

File: Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Web/Controllers/NotificationsController.cs [hunk 4]
```diff
@@ -96,5 +110,142 @@ public ActionResult RenderingTemplate([FromBody]NotificationTemplateRequest requ
...
+        /// <remarks>
+        /// Method returns notification journal page with array of notification, that was send, sending or will be send in future. Result contains total count, that can be used
+        /// for paging.
+        /// </remarks>
+        /// <param name="start">Page setting start</param>
+        /// <param name="count">Page setting count</param>
+        /// <param name="sort">Sort expression</param>
+        [HttpGet]
+        [Route("journal")]
+        public async Task<ActionResult<NotificationMessageSearchResult>> GetNotificationJournal([FromQuery]int start, [FromQuery]int count, [FromQuery]string sort)
```
Comment (tatarincev) [inline Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Web/Controllers/NotificationsController.cs]: Need to replace multiple formal parameters to criteria

File: Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Web/Controllers/NotificationsController.cs [hunk 5]
```diff
@@ -96,5 +110,142 @@ public ActionResult RenderingTemplate([FromBody]NotificationTemplateRequest requ
...
+        {
+            notification.TenantIdentity = new TenantIdentity(request.ObjectId, request.ObjectTypeId);
+
+            foreach (var parameter in request.NotificationParameters)
+            {
+                SetValue(notification, parameter);
+            }
+        }
+
+        private void SetValue(Notification notification, NotificationParameter param)
```
Comment (tatarincev) [inline Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Web/Controllers/NotificationsController.cs]: Could you move out  this method from controller to extensions?

File: Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Web/Controllers/NotificationsController.cs [hunk 6]
```diff
@@ -96,5 +110,142 @@ public ActionResult RenderingTemplate([FromBody]NotificationTemplateRequest requ
...
+        /// </summary>
+        /// <remarks>
+        /// Method sending notification, that based on notification template. Template for rendering chosen by type, objectId, objectTypeId, language.
+        /// Parameters for template may be prepared by the method of getTestingParameters. Method returns string. If sending finished with success status
+        /// this string is empty, otherwise string contains error message.
+        /// </remarks>
+        /// <param name="request">Notification request</param>
+        [HttpPost]
+        [Route("template/sendnotification")]
+        public async Task<ActionResult<NotificationSendResult>> SendNotification([FromBody]NotificationRequest request)
```
Comment (tatarincev) [inline Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Web/Controllers/NotificationsController.cs]: Mark this method by special comment what this API left for backward compatibility and can be replaced  to other. (need to create a new method that allows to send any  notification by API call)

File: Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Data/Services/NotificationSearchService.cs (hunk 3 above)
Comment (kostyrin) [reply Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Data/Services/NotificationSearchService.cs]: image

File: Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Core/Model/Search/NotificationMessageSearchCriteria.cs (hunk 1 above)
Comment (kostyrin) [reply Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Core/Model/Search/NotificationMessageSearchCriteria.cs]: actually, don't need it. have 
`function getSearchCriteria()
        {
            var searchCriteria = {
                catalogId: blade.catalogId,
                categoryId: blade.categoryId,
                keyword: filter.keyword ? filter.keyword : undefined,
                sort: uiGridHelper.getSortExpression($scope),
                skip: ($scope.pageSettings.currentPage - 1) * $scope.pageSettings.itemsPerPageCount,
                take: $scope.pageSettings.itemsPerPageCount
            };
            return searchCriteria;
        }`

File: Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Web/Controllers/NotificationsController.cs (hunk 6 above)
Comment (kostyrin) [reply Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Web/Controllers/NotificationsController.cs]: mark as [Obsolete]

**Pull request merged**